### PR TITLE
autoconfig: automatically starts the docker listener

### PR DESCRIPTION
**Pull request by JulienBalestra**
### What does this PR do?

This PR allows to select the listener name **auto** like:
```yaml
listeners:
  - name: auto
```

The listener **auto** adds the docker listener if it's not already specified in the **listeners** and if the host can run it.

To determine if the host can run it, we call `docker.GetDockerUtil` and observe an error equal to `nil`.

At the moment, there is only one auto-listener: **docker**.

### CODE REVIEW

File: pkg/config/config.go [hunk 1]
```diff
@@ -303,3 +304,35 @@ func IsContainerized() bool {
 func FileUsedDir() string {
 	return filepath.Dir(Datadog.ConfigFileUsed())
 }
+
+func isDockerSocket() bool {
```
Comment (mfpierre) [inline pkg/config/config.go]: `Mounted` or `Present` at the end of the function name would sound nicer

**Comment by xvello** [comment]
I'm not a fan of having that kind of detection logic out of the `util/docker` package. Plus, we might want to use the `kubelet` listener instead, or even disable AD altogether (think about a gitlab runner host).

I'd rather:

- have an explicit `auto` setting, that might become the default:
```
listeners:
  - name: auto
```
- reuse the `Retrier` interface instead of creating new detection code, see https://github.com/DataDog/datadog-agent/pull/861 for how the tagger implements it

File: pkg/config/config.go [hunk 2]
```diff
@@ -303,3 +304,35 @@ func IsContainerized() bool {
...
+// AutoAddDockerListener adds the docker listener if the docker socket is found
+func AutoAddDockerListener(listeners []Listeners) []Listeners {
+	isDocker := false
+	for _, l := range listeners {
+		if l.Name == "docker" {
+			isDocker = true
+		}
+	}
+	if isDocker == false && isDockerSocket() == true {
+		log.Infof("docker socket found: add docker listener")
```
Comment (mfpierre) [inline pkg/config/config.go]: s/add/adding

File: pkg/config/config_template.yaml [hunk 3]
```diff
@@ -204,6 +204,8 @@ metadata_collectors:
 #
 # container_proc_root: /host/proc
 #
+# Choose "auto" if you want to let the agent find any relevant listener on your host
```
Comment (masci) [inline pkg/config/config_template.yaml]: Can we make this more clear with an example showing how to choose "auto"?

File: pkg/config/auto_listener.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import (
+	"os"
+	"path"
+
+	log "github.com/cihub/seelog"
+)
+
+// AutoAddListeners adds the docker listener if needed
+// TODO: support more listeners (kubelet, ...)
+func AutoAddListeners(listeners []Listeners) []Listeners {
```
Comment (masci) [inline pkg/config/auto_listener.go]: The semantic is not clear to me: what happens if I select "auto" and "docker"? Or, in the future, "auto" and "kubelet"?

File: pkg/config/auto_listener.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	return false
+}
+
+// isDockerHostEnv checks if the DOCKER_HOST environment variable is set
+func isDockerHostEnv() bool {
+	return os.Getenv("DOCKER_HOST") != ""
+}
+
+// isDockerRun check with several options to determine if docker runs
+func isDockerRun() bool {
```
Comment (masci) [inline pkg/config/auto_listener.go]: I'd call this `isDockerRunning`

**Review by masci — CHANGES_REQUESTED** [review]
Left a couple of comments but I might miss the point of this changes, can you please elaborate the use case and the solution implemented in the PR description?

**Review by hkaj — CHANGES_REQUESTED** [review]
Looks good, just need to update the headers for the new year (🎉 ) and let's merge it.

**Comment by JulienBalestra** [comment]
@hkaj I rebased and updated the dates :+1:

**Review by hkaj — APPROVED** [review]
🚢 it

**Pull request merged**